### ISSUE TITLE
Rule Proposal: no-ascii

### ISSUE THREAD

**Issue by sigorilla** [eslint/eslint#6341]
Hi everyone. One day my teammate wrote object key with non-ASCII characters. It led to problems with renaming. So I want proposal new rule `no-ascii`.

**When does this rule warn? Please describe and show example code:**

```js
// Russian letter `Й`.
var obj = {
  Й: 'value'
};

// Japanese variable.
var 日 = 'Japanese';
日.length; // 8
```

**Is this rule preventing an error or is it stylistic?**

In JS there isn't prohibited non-ASCII characters (it works correctly), but in some projects this is disallowed.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

I think it's global rule for everyone.

**Are you willing to create the rule yourself?**

I found [solution](https://gist.github.com/fand/c13213f81bfce11f4132) by @fand, but I made [some changes](https://gist.github.com/sigorilla/5a9867ff995b76990d4512f70ae7bf43).

**Comment by vitorbal** [comment]
Hi @sigorilla , thanks for the suggestion! Few notes to help us consider this rule:

- I don't think `no-ascii` is a good name, it sounds like the rule is to forbid ascii characters, which is the opposite of what it does in reality.
- Even though stylistic rules are quite subjective, I think it would be helpful to expand on your arguments on why non-ASCII characters should be disallowed. For example, you mentioned problems with renaming. What kind of problems?
- Could you add some more examples of when this rule would pass/fail, e.g. non-ascii characters in Literals, TemplateLiterals, etc.

**Comment by sigorilla** [comment]
Thank you for your feedback!

> I don't think no-ascii is a good name, it sounds like the rule is to forbid ascii characters, which is the opposite of what it does in reality.

What about `variable-without-non-ascii`? Or `without-non-ascii`?

> What kind of problems?

I started rename all variables with non-ASCII letters, so I began to look for all occurrences of this characters in the whole project. It's not fast. And you have to be very careful not to miss something.

> Could you add some more examples of when this rule would pass/fail

### Pass

**Literal**

```js
var foo = 'строка';
var bar = {
    baz: 'string ' + foo
};
```

**TemplateLiteral**

```js
var number = 1;
var string1 = `value: ${number + 1}`;
var string2 = `значение: ${number + 2}`;
```

**RegularExpression**

```js
var reg1 = new RegExp('[а-яА-Я]+', 'g');
var reg2 = /[^ё]+/gi;
```

### Fail

All occurrences of `Identifier` (or `JSXIdentifier`).

```jsx
// Literal
var число = 1;
// -^^^^^-----
var vaр = 2;
// ---^-----

// TemplateLiteral
var строка = `Значение: ${число + vaр}`;
// -^^^^^^----------------^^^^^-----^---

// ObjectExpression
var объект = {строка: строка, число};
// -^^^^^^----^^^^^^--^^^^^^--^^^^^--
// ArrayExpression
var массив = [объект, объект];
// -^^^^^^----^^^^^^--^^^^^^--

// JSXElement
var Блок = (<div>{props.текст}</div>);
// -^^^^----------------^^^^^---------

// JSXAttribute
var Результат = (<Блок текст={объект.строка} />);
// -^^^^^^^^^-----^^^^-^^^^^--^^^^^^-^^^^^^------

// FunctionExpression
var функция = function (аргумент) {
// -^^^^^^^-------------^^^^^^^^---
    return аргумент;
// --------^^^^^^^^-
};
```

**Comment by pedrottimark** [comment]
@sigorilla Welcome to ESLint!

As you think about what to call this rule and how to describe its goal, here are a few existing rules about names and character codes that might be helpful to read about:
* id-blacklist
* id-length
* id-match
* no-irregular-whitespace

**Comment by ilyavolodin** [comment]
I think this can already be accomplished with `id-match` rule with the following configuration:

```js
"id-match": ["error", "^[a-zA-Z]*$"]
```

**Comment by pedrottimark** [comment]
@ilyavolodin Do you think I should write a sentence about this use case for `id-match` rule doc?

**Comment by ilyavolodin** [comment]
I'm not sure that's special case. RegEx are very powerful and allow you to cover a lot of use-cases. We will not be able to mention all of them.

**Comment by sigorilla** [comment]
Oh, thank you @ilyavolodin. That's what I need!
